### PR TITLE
PLAT-1110 Implement WHEEL event

### PR DESCRIPTION
**Pull request by alexander-schmidt**
Peek 2022-08-26 11-21

### CODE REVIEW

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts [hunk 1]
```diff
@@ -120,6 +134,8 @@ class AjaxRequestMessage {
 				return false; // allow new passive events, i.e. implicitly triggered events
 			} else if(this.isKeyEventType()) {
 				return false; // never drop key events
+			} else if(this.isWheelType()) {
```
Comment (oliver-richers) [inline platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts:137]:
```suggestion
			} else if(this.isWheelEventType()) {
```

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts [hunk 2]
```diff
@@ -128,6 +144,22 @@ class AjaxRequestMessage {
 		}
 	}
 
+	private isSameDeltaDirections(other: AjaxRequestMessage) {
```
Comment (oliver-richers) [inline platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts:147]: i think the type definitions are useless, and isn't there a more explicit way to convert to number than to use the unary `+` operator?

File: platform-dom/src/main/java/com/softicar/platform/dom/event/IDomEvent.java [hunk 3]
```diff
@@ -118,6 +118,33 @@ public interface IDomEvent {
...
 	DomRect getBoundingClientRect();
 
+	/**
+	 * Returns the delta-x value of a {@link DomEventType#WHEEL} event.
+	 * <p>
+	 * Returns 0 for any other {@link DomEventType}.
+	 *
+	 * @return the delta-x value (never <i>null</i>)
+	 */
+	Double getDeltaX();
```
Comment (oliver-richers) [inline platform-dom/src/main/java/com/softicar/platform/dom/event/IDomEvent.java:128]: why support nullability when we don't need it?
```suggestion
	double getDeltaX();
```

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts (hunk 1 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts:137]: oops

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts (hunk 2 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts:147]: there is. we can just use `Number(someString)` instead.

File: platform-dom/src/main/java/com/softicar/platform/dom/event/IDomEvent.java (hunk 3 above)
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/event/IDomEvent.java:128]: sure that

File: platform-core-module/src/main/java/com/softicar/platform/core/module/start/page/StartPageDiv.java [hunk 4]
```diff
@@ -32,6 +37,33 @@ public StartPageDiv() {
 		appendSectionsSeparatedByHr(sections);
 
 		addMarker(CoreTestMarker.START_PAGE_MAIN_ELEMENT);
+
+		// FIXME remove
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/start/page/StartPageDiv.java:41]: so do it :)

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts [hunk 5]
```diff
@@ -128,6 +144,22 @@ class AjaxRequestMessage {
...
+		let thisDeltaX = Number(this.data.get('deltaX') ?? 0);
+		let otherDeltaX = Number(other.data.get('deltaX') ?? 0);
+		let deltaXSameSign = (thisDeltaX == 0 && otherDeltaX == 0) || thisDeltaX * otherDeltaX > 0;
```
Comment (oliver-richers) [inline platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts:150]: how about
```suggestion
		let thisDeltaX = Number(this.data.get('deltaX') ?? 0);
		let otherDeltaX = Number(other.data.get('deltaX') ?? 0);
		let deltaXSameSign = Math.sign(thisDeltaX) == Math.sign(otherDeltaX);
```

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts (hunk 5 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/request/AjaxRequestMessage.ts:150]: looks fine, assuming that the distinction between positive and negative zero makes no difference here. cf https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/sign

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineEventSimulator.java [hunk 6]
```diff
@@ -47,6 +47,23 @@ public void simulateInput(IDomNode node) {
 		javascriptExecutor.execute(javascript, engine.getOutput().getAttributeValue(node, "id"));
 	}
 
+	@Override
+	public void simulateWheel(IDomNode node, int deltaY) {
```
Comment (oliver-richers) [inline platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineEventSimulator.java:51]: why do we need to simulate this?

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java [hunk 7]
```diff
@@ -756,4 +780,9 @@ default void simulateInput(IDomNode target) {
 
 		getTestEngine().getEventSimulator().simulateInput(target);
 	}
+
+	default void simulateWheel(IDomNode node, int deltaY) {
```
Comment (oliver-richers) [inline platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java:784]: what is the unit of `deltaY`?

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java [hunk 8]
```diff
@@ -230,6 +230,30 @@ default void mouseUpUnsafe() {
...
+	/**
+	 * Rotates the mouse wheel downwards.
+	 *
+	 * @param node
+	 *            the {@link IDomNode} to rotate the mouse wheel on (never
+	 *            <i>null</i>)
+	 */
+	default void mouseWheelDown(IDomNode node) {
+
+		simulateWheel(node, 200);
```
Comment (oliver-richers) [inline platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java:242]: magic constant `200`?

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java [hunk 9]
```diff
@@ -230,6 +230,30 @@ default void mouseUpUnsafe() {
...
+	/**
+	 * Rotates the mouse wheel upwards.
+	 *
+	 * @param node
+	 *            the {@link IDomNode} to rotate the mouse wheel on (never
+	 *            <i>null</i>)
+	 */
+	default void mouseWheelUp(IDomNode node) {
+
+		simulateWheel(node, -200);
```
Comment (oliver-richers) [inline platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java:254]: magic constant `-200`?

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+		mouseWheelDown(testDiv);
+		waitForServer();
+
+		var events = testDiv.getEvents();
+		new WheelEventAsserter(events)//
+			.nextEvent()
+			.assertDeltaXZero()
+			.assertDeltaYPositive()
+			.assertDeltaZZero()
```
Comment (oliver-richers) [inline platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:29]: shouldn't we simplify this to `assertWheelDown`?

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+		mouseWheelUp(testDiv);
+		waitForServer();
+
+		var events = testDiv.getEvents();
+		new WheelEventAsserter(events)//
+			.nextEvent()
+			.assertDeltaXZero()
+			.assertDeltaYNegative()
+			.assertDeltaZZero()
```
Comment (oliver-richers) [inline platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:46]: shouldn't we simplify this to `assertWheelUp`?

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+		mouseWheelDown(testDiv);
+		waitForServer();
+
+		var events = testDiv.getEvents();
+		new WheelEventAsserter(events)//
+
+			.nextEvent()
+			.assertDeltaXZero()
+			.assertDeltaYPositive()
+			.assertDeltaZZero()
```
Comment (oliver-richers) [inline platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:71]: simplify, see above

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+			.nextEvent()
+			.assertDeltaXZero()
+			.assertDeltaYNegative()
+			.assertDeltaZZero()
```
Comment (oliver-richers) [inline platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:76]: simplify, see above

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineEventSimulator.java (hunk 6 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineEventSimulator.java:51]: It does not. While wheel input was generally implemented with Selenium 4.2, the API thus far only allows for scrolling of the viewport / document.

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java (hunk 8 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java:242]: extracted a constant

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java (hunk 9 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java:254]: extracted a constant

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java (hunk 10 above)
Comment (alexander-schmidt) [reply platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:29]: yep

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java (hunk 11 above)
Comment (alexander-schmidt) [reply platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:46]: yep

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java (hunk 12 above)
Comment (alexander-schmidt) [reply platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:71]: yep

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java (hunk 13 above)
Comment (alexander-schmidt) [reply platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/wheel/AjaxDomWheelEventTest.java:76]: yep

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineEventSimulator.java (hunk 6 above)
Comment (oliver-richers) [reply platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineEventSimulator.java:51]: indeed, suprising

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java (hunk 7 above)
Comment (oliver-richers) [reply platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java:786]: we might add that as Javadoc, what you think?

**Review by oliver-richers — DISMISSED** [review]

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java (hunk 7 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/interfaces/IAjaxSeleniumLowLevelTestEngineMethods.java:786]: did that. also found out that px is the default.

**Pull request merged**